### PR TITLE
AP_VisualOdom: NoopLoop driver

### PR DESCRIPTION
**Pull request by chobitsfan**
This is based on #14504 by @rmackay9. I only make a little changes

1. The nooploop beacons must be setup so that the system's positive Y axis point to North
noonloop
2. According to developers in nooploop, output velocity data is computed from kalman filter fusing onboard IMU and UWB position. I think there may be not much help to feed this velocity data to Ardupilot

flight data
[35 1980-1-1 08-00-00.zip](https://github.com/ArduPilot/ardupilot/files/5205546/35.1980-1-1.08-00-00.zip)

### CODE REVIEW

File: libraries/AP_VisualOdom/AP_VisualOdom_NoopLoop.cpp [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+    const float att_err = _frontend.get_yaw_noise();
+    const uint8_t reset_counter = 0;
+
+    // write data to EKF
+    AP::ahrs().writeExtNavData(pos_m, att_none, pos_err, att_err, now_ms, _frontend.get_delay_ms(), reset_counter);
+
+    // log sensor data
+    AP::logger().Write_VisualPosition(remote_systime_us, now_ms, pos_m.x, pos_m.y, pos_m.z, 0.0f, 0.0f, 0.0f, pos_err, att_err, reset_counter);
+
+    // debug
```
Comment (rmackay9) [inline libraries/AP_VisualOdom/AP_VisualOdom_NoopLoop.cpp:181]: Let's remove this debug

File: libraries/AP_VisualOdom/AP_VisualOdom_NoopLoop.cpp [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+                    (double)vel_ms.z);
+    }
+    */
+}
+
+// returns false if we fail arming checks, in which case the buffer will be populated with a failure message
+bool AP_VisualOdom_NoopLoop::pre_arm_check(char *failure_msg, uint8_t failure_msg_len) const
+{
+    if (_uart == nullptr) {
+        hal.util->snprintf(failure_msg, failure_msg_len, "No SERIALx_PROTOCOL = 29");
```
Comment (rmackay9) [inline libraries/AP_VisualOdom/AP_VisualOdom_NoopLoop.cpp:203]: we should change, "29" to "34"

File: libraries/SITL/SIM_Vicon.h [hunk 3]
```diff
@@ -13,7 +13,22 @@
...
 /*
-  simple particle sensor simulation
+  simple Vicon simulation
+
+  NoopLoop testing:
+
+./Tools/autotest/sim_vehicle.py --gdb --debug -v ArduCopter -A --uartF=sim:vicon --speedup=1
+
+param set GPS_TYPE 0
+param set EK2_GPS_TYPE 3
```
Comment (rmackay9) [inline libraries/SITL/SIM_Vicon.h:23]: I'm not sure it's a good idea to have these param values here.  It's probably better instead to have them on the wiki "Using SITL" page.  In any case I think we recommend using EKF3 these days.

File: libraries/SITL/SIM_Vicon.h [hunk 4]
```diff
@@ -52,14 +67,16 @@ class Vicon : public SerialDevice {
     // buffer of messages to send
     struct {
         uint64_t time_send_us;      // system time this message should be sent or 0 if no message to send
-        mavlink_message_t obs_msg;  // message to be sent
+        uint8_t obs_msg[300];  // message to be sent
```
Comment (rmackay9) [inline libraries/SITL/SIM_Vicon.h:70]: let's align the comments with the line above.

File: libraries/SITL/SIM_Vicon.cpp [hunk 5]
```diff
@@ -169,16 +172,16 @@ void Vicon::update_vicon_position_estimate(const Location &loc,
             pitch,
             yaw,
             NULL, 0);
-        msg_buf[msg_buf_index].time_send_us = time_send_us;
+        have_msg = true;
```
Comment (rmackay9) [inline libraries/SITL/SIM_Vicon.cpp:175]: I'm a little worried that this change is removing the ability to add delay into the messages.  Does it mean that we no longer use, "time_send_us"?  can this variable be removed?

**Comment by rmackay9** [comment]
@chobitsfan, TBH, one of the reasons I didn't push this further was because of the restriction on how the beacons must be arranged.  I had been thinking that a better implementation would be to add support for NoopLoop to AP_Beacon.  I'm not against this going in because it adds value but I suspect we are going to get a lot of support calls because many users won't be able to arrange the beacons exactly North.

@peterbarker, @khancyr, could one of you review this?

**Comment by rmackay9** [comment]
@chobitsfan, sorry, instead of me dancing around the issue, let me be more clear :-).... you wouldn't want to try to implement NoopLoop support in AP_Beacon instead would you?  No pressure of course.. your testing and fixes are greatly appreciated!

**Comment by chobitsfan** [comment]
Thank you very much @rmackay9 

> I had been thinking that a better implementation would be to add support for NoopLoop to AP_Beacon

Yes, I agree with you. I could try to do it because I need this feature, too.

**Comment by chobitsfan** [comment]
Hi @rmackay9 May I ask a question about AP_Beacon library here?
What is order of beacon_instance in https://github.com/ArduPilot/ardupilot/blob/f364dc933d060eb8e631dfd5ff11ab6769a88a2b/libraries/AP_Beacon/AP_Beacon_Backend.h#L44
Did it follow the order of pozyx beacon?  Thank you very much
pozyx-anchor-layout

**Comment by chobitsfan** [comment]
replaced by #15372

**Comment by YANG-DEMIN** [comment]
> Thank you very much @rmackay9
> 
> > I had been thinking that a better implementation would be to add support for NoopLoop to AP_Beacon
> 
> Yes, I agree with you. I could try to do it because I need this feature, too.

Hello! I'm using the Linktrack module to replace gps.
My hardware is CUAV V5 nano and Linktrack LPS.
I have tried 2 methods:
1. Changing the UWB protocol to NEMA;
    Using PX4 Firmware, and replace the GPS with Linktrack in the same place;
    But it was not recognized by FCU as GPS, I find that PX4 cant support the NEMA protocol.
    So i turn to the secod method
2. Setting as (https://ardupilot.org/copter/docs/common-nooploop.html)
    Using the master Firmware of CUAV V5 Nano, but i cant find some parameter, such as BCN_TYPE(only have 0,1,2,10 don't 
     have the nooploop chioce), EKF3_BCN_DELAY.
I have confused for 2 days. Can you have some solutions?
Thank you very much!
      yang

**Comment by chobitsfan** [comment]
Hi @YANG-DEMIN 
> i cant find some parameter, such as BCN_TYPE(only have 0,1,2,10 don't have the nooploop chioce), 

MissionPlanner may not update its descriptions. You can just set BCN_TYPE to 3

> EKF3_BCN_DELAY.

Did you set EK3_ENABLE to 1 and reboot? You may need to set AHRS_EKF_TYPE to 3

**Comment by YANG-DEMIN** [comment]
> Hi @YANG-DEMIN
> 
> > i cant find some parameter, such as BCN_TYPE(only have 0,1,2,10 don't have the nooploop chioce),
> 
> MissionPlanner may not update its descriptions. You can just set BCN_TYPE to 3
> 
> > EKF3_BCN_DELAY.
> 
> Did you set EK3_ENABLE to 1 and reboot? You may need to set AHRS_EKF_TYPE to 3

Hi, Thank you very much. As your instruction, I fix the para problem and change them to desired value in MP.Then, I turn to QGC(because i am used to it).
But, I still can not see what I want. There are a warning as `PreArm: AHRS: EKF3 sources require GPS.`

I think I still your help. Thanks for you agin.
There are some setting of my FCU:

qgc
qgc2
serial1
linktrack

**Comment by chobitsfan** [comment]
Hi @YANG-DEMIN 

You may need to set EK3_GPS_CHECK = 0 and EK3_GPS_TYPE = 3 (or EK3_SRC1_POSXY = 4, EK3_SRC1_VELXY = EK3_SRC1_VELZ = 0 if you use 4.1.0-dev)

Maybe we could continue this discussion in https://discuss.ardupilot.org/

**Comment by YANG-DEMIN** [comment]
> 嗨@ YANG-DEMIN
> 
> 您可能需要设置EK3_GPS_CHECK = 0和EK3_GPS_TYPE = 3（或EK3_SRC1_POSXY = 4，如果使用4.1.0-dev，则EK3_SRC1_VELXY = EK3_SRC1_VELZ = 0）
> 
> 也许我们可以在[https://discuss.ardupilot.org/中](https://discuss.ardupilot.org/)继续讨论[](https://discuss.ardupilot.org/)

(https://discuss.ardupilot.org/t/nooploop-linktrack-setting-on-ardupilot-v5-copter-4-0dev/65132)
Here it is ! Thanks for you reply and i get some progress.

**Comment by rmackay9** [comment]
Thanks for pitching in @chobitsfan.  I think once we switch master to EKF3 (https://github.com/ArduPilot/ardupilot/pull/15909) we will probably need to update many wiki pages to reflect the new parameter names.  We might need your help on that but I suspect that Henry and I will handle most of that.  Thanks again!

**Comment by chobitsfan** [comment]
>We might need your help on that but I suspect that Henry and I will handle most of that. Thanks again!

Hi @rmackay9 
I am happy to help.

**Comment by zhuyang1** [comment]
https://discuss.ardupilot.org/t/pix-mini-4-1dev-nooploop-linktrack-set/77132
I also encountered this problem, can't solve, I tried many ways can't identify the nooploop

**Pull request closed without merging**